README.md

# Tailscale Lambda Extension for Containers

Inspired by Corey Quinn's [tailscale-layer](https://github.com/QuinnyPig/tailscale-layer) project. This project aims to do the same thing but usable for those working with any Lambda Container runtime language. Our extension has been modified from Corey's upstream work. I have removed verbose logging and event lifecycle hooks. This has reduced startup overhead, duplicate tailscale up calls, and more configuration options.

## Installation

We have pre-build container packages which include the Tailscale Lambda Extension within a basic Docker scratch image containing everything needed to copy to your `/opt` directory. For example:

```dockerfile
FROM public.ecr.aws/lambda/ruby:3.2
RUN yum install -y curl
COPY --from=ghcr.io/rails-lambda/tailscale-extension-amzn:1 /opt /opt
```

⚠️ Make sure `curl` is installed since the Tailscale Extension script will need it.

### Environment Variables

You will need to generate an [ephemeral key](https://login.tailscale.com/admin/settings/authkeys) within your Tailscale account. Note, this will expire and need updating depending on the time period you select. List of the environment variables include:

- `TS_KEY` - Required. Your ephemeral key.
- `TS_HOSTNAME` - Optional. The value of `--hostname` parameter. Default `lambda`.

### OSs & Architectures

We publish multi-platform images for both `linux/amd64` and `linux/arm64/v8` and we also have Amazon Linux 2 and Debian/Ubuntu variants.

- ghcr.io/rails-lambda/tailscale-extension-amzn
- ghcr.io/rails-lambda/tailscale-extension-debian

## Example Usage

Once your Lambda function starts, you will have a [SOCKS5](https://en.wikipedia.org/wiki/SOCKS) proxy which can communicate with your Tailscale tailnet at `http://localhost:1055`. Here is an example of how to leverage that with Ruby's [socksify](https://github.com/astro/socksify-ruby) gem.

```ruby
Net::HTTP.socks_proxy('localhost', 1055).start(...) do |http|
  # your http code here...
end
```

Again, this extension is not coupled to any runtime language. So how you use the SOCKS5 proxy is up to you. Enjoy!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0] - 2023-05-25

### Added

- New container based install for and initial release.



src/app.mjs
import * as http from "http";
export const lambdaHandler = async (event, context) => {
  const payload = JSON.stringify({ event: event, context: context });
  const options = {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Content-Length": payload.length,
    },
    hostname: process.env.TS_REMOTE_PROXY_HOST,
    port: Number(process.env.TS_REMOTE_PROXY_PORT),
    path: "/",
  };
  const proxyResponse = await new Promise((resolve, reject) => {
    const req = http.request(options, (res) => {
      let body = "";
      res.on("data", (chunk) => (body += chunk));
      res.on("end", () => resolve(JSON.parse(body)));
    });
    req.setTimeout(20000, () => {
      req.destroy();
      reject(new Error("Request timed out"));
    });
    req.write(payload);
    req.end();
  });
  return JSON.parse(proxyResponse);
};


src/tailscale.sh
#!/bin/bash
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

set -euo pipefail
set +x

# Configuration
OWN_FILENAME="$(basename "$0")"
LAMBDA_EXTENSION_NAME="$OWN_FILENAME" # (external) extension name has to match the filename
TMPFILE="/tmp/tailscale.data"

# Registration
HEADERS="$(mktemp)"
echo "[${LAMBDA_EXTENSION_NAME}] Registering at http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register"
curl -sS -LD "$HEADERS" -XPOST "http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register" --header "Lambda-Extension-Name: ${LAMBDA_EXTENSION_NAME}" -d "{ \"events\": [\"SHUTDOWN\"]}" > $TMPFILE

# Extract Extension ID from response headers
RESPONSE=$(<$TMPFILE)
EXTENSION_ID=$(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)
echo "[${LAMBDA_EXTENSION_NAME}] Registration response: ${RESPONSE} with EXTENSION_ID $(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)"

# Start the Tailscale process
echo "[${LAMBDA_EXTENSION_NAME}] Tailscale process..." 1>&2;
/opt/bin/tailscaled --tun=userspace-networking --socks5-server=localhost:1055 --socket=/tmp/tailscale.sock --state=/tmp/tailscale &
TAILSCALED_PID=$!
echo "[${LAMBDA_EXTENSION_NAME}] TAILSCALED_PID: ${TAILSCALED_PID}" 1>&2;
sleep 1

# Tailscale up
echo "[${LAMBDA_EXTENSION_NAME}] Tailscale up..." 1>&2;
TS_HOSTNAME=${TS_HOSTNAME:-lambda}
if [ "${AWS_LAMBDA_FUNCTION_VERSION}" != '$LATEST' ]; then
  TS_HOSTNAME="${TS_HOSTNAME}-v${AWS_LAMBDA_FUNCTION_VERSION}"
fi
until /opt/bin/tailscale --socket=/tmp/tailscale.sock up --authkey="${TS_KEY}" --hostname="${TS_HOSTNAME}"
do
  sleep 0.1
done
sleep 1

# Waiting for SHUTDOWN event.
while true
do
  echo "[${LAMBDA_EXTENSION_NAME}] Waiting for event. Get /next event from http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/event/next"
  # Get an event. The HTTP request will block until one is received
  curl -sS -L -XGET "http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/event/next" --header "Lambda-Extension-Identifier: ${EXTENSION_ID}" > $TMPFILE
  echo "[${LAMBDA_EXTENSION_NAME}] Event received. Processing..."  1>&2;
  EVENT_DATA=$(<$TMPFILE)
  cat $EVENT_DATA 1>&2;
  echo "[${LAMBDA_EXTENSION_NAME}] Calling tailscale down..."  1>&2;
  /opt/bin/tailscale --socket=/tmp/tailscale.sock down
  echo "[${LAMBDA_EXTENSION_NAME}] Sending term to ${TAILSCALED_PID}..."  1>&2;
  kill -TERM "$TAILSCALED_PID" 2>/dev/null
  exit 0
done


src/live-development-proxy.sh
#!/bin/bash
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

set -euo pipefail
set +x

OWN_FILENAME="$(basename $0)"
LAMBDA_EXTENSION_NAME="$OWN_FILENAME" # (external) extension name has to match the filename
TMPFILE=/tmp/$OWN_FILENAME

# Graceful Shutdown
_term() {
  echo "[${LAMBDA_EXTENSION_NAME}] Received SIGTERM"
  # forward SIGTERM to child procs and exit
  kill -TERM "$PID" 2>/dev/null
  echo "[${LAMBDA_EXTENSION_NAME}] Exiting"
  exit 0
}

forward_sigterm_and_wait() {
  trap _term SIGTERM
  wait "$PID"
  trap - SIGTERM
}

# Registration
HEADERS="$(mktemp)"
echo "[${LAMBDA_EXTENSION_NAME}] Registering at http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register"
curl -sS -LD "$HEADERS" -XPOST "http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register" --header "Lambda-Extension-Name: ${LAMBDA_EXTENSION_NAME}" -d "{ \"events\": [\"SHUTDOWN\", \"INVOKE\"]}" > $TMPFILE

RESPONSE=$(<$TMPFILE)
HEADINFO=$(<$HEADERS)
# Extract Extension ID from response headers
EXTENSION_ID=$(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)
echo "[${LAMBDA_EXTENSION_NAME}] Registration response: ${RESPONSE} with EXTENSION_ID $(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)"

# Start the Tailscale process
# /opt/bin/tailscaled --tun=userspace-networking --socks5-server=localhost:1055 --socket=/tmp/tailscale.sock --state /tmp/tailscale &
# until /opt/bin/tailscale --socket=/tmp/tailscale.sock up --authkey=$TS_KEY
# do
#   sleep 0.1
# done

TS_HOSTNAME=${TS_HOSTNAME:-lambda}
/opt/bin/tailscaled --tun=userspace-networking --socks5-server=localhost:1055 & sleep 2 &
/opt/bin/tailscale up --authkey=${TS_KEY} --hostname=${TS_HOSTNAME}

# Event processing
while true
do
  echo "[${LAMBDA_EXTENSION_NAME}] Waiting for event. Get /next event from http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/event/next"
  # Get an event. The HTTP request will block until one is received
  curl -sS -L -XGET "http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/event/next" --header "Lambda-Extension-Identifier: ${EXTENSION_ID}" > $TMPFILE &
  PID=$!
  forward_sigterm_and_wait
  if [[ $EVENT_DATA == *"SHUTDOWN"* ]]; then
    echo "[extension: ${LAMBDA_EXTENSION_NAME}] Received SHUTDOWN event. Exiting."  1>&2;
    # Cleanly shut down the Tailscale process
    /opt/bin/tailscale --socket=/tmp/tailscale.sock down
    exit 0 # Exit if we receive a SHUTDOWN event
  fi
  sleep 1
done
